### PR TITLE
C19 Adagrams Divya Naidoo

### PR DESCRIPTION
**Pull request by dnai55**

### CODE REVIEW

File: adagrams/game.py [hunk 1]
```diff
@@ -1,11 +1,139 @@
...
+        'V': 2, 
+        'W': 2, 
+        'X': 1, 
+        'Y': 2, 
+        'Z': 1
+    }
+
+    letter_bank = []
+    for i in range(10):
+        letter = random.choice(list(LETTER_POOL.keys()))
```
Comment (mmcknett-ada) [inline adagrams/game.py:35]: This way of choosing letters randomly doesn't quite meet the requirements. Notice that your solution is just as likely to draw a `Z` as it is to draw an `E`. The requirements say, "Since there are 12 `E`s but only 1 `Z`, it should be 12 times as likely for the user to draw an `E` as a `Z`". It turns out, our tests don’t test for that. (That test is quite hard to write!) Think about how you might modify your code to fulfill this requirement.

Regarding `random.choice`: when bringing in functions from imported modules (that we haven’t covered in the curriculum), please consider including a comment about how you understand the function to work, and how you researched it. Now that we’ve discussed Big O, it’s also useful to think about what the time/space complexity of the function might be. Thinking about how we can implement this functionality ourselves can help us gain insight into what that might be.

Remember that not every language has the same built-in functions, so we should be able to implement this same functionality ourselves. In the projects, we don’t ask for any features that we don’t think you should be able to write yourself. For drawing a random hand, the only external function that would be “required” is a way to pick a random number (such as randint). At this stage in our coding journeys, it’s often more valuable to re-implement things by hand to practice thinking about how to break down and approach the problem than to use a single library function to solve the problem for us.

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,139 @@
...
+        letter_bank.append(letter)
+        LETTER_POOL[letter] -= 1
+        
+        if LETTER_POOL[letter] == 0:
+            LETTER_POOL.pop(letter)
```
Comment (mmcknett-ada) [inline adagrams/game.py:40]: This was a good way to handle looping exactly 10 times and dealing with the 0 case.

[nit] Since you don't use the result of `pop`, you can use `del` instead:

```python
        if LETTER_POOL[letter] == 0:
            del LETTER_POOL[letter]
```

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,139 @@
...
+    'R': 1, 
+    'S': 1, 
+    'T': 1, 
+    'U': 1, 
+    'V': 4, 
+    'W': 4, 
+    'X': 8, 
+    'Y': 4, 
+    'Z': 10
+}
```
Comment (mmcknett-ada) [inline adagrams/game.py:85]: [nit] Please indent

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,139 @@
...
+    score = 0
+    for letter in word.upper():
+        point_value = LETTER_POINT_VALUE[letter]
+        score += point_value
```
Comment (mmcknett-ada) [inline adagrams/game.py:89]: [discussion] I like intermediate values like `point_value` sometimes, and sometimes I elide them. For me, it comes down to readability and how often I think I'm going to debug the line. I think what you have is valid; if you thought it were more readable, it would be equally valid to do on one line:

```python
score += LETTER_POINT_VALUE[letter]
```

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,139 @@
...
+    if len(tie_words) == 1:
+        return max_key, max_value
+    
+    word_lengths = {}
+    for word in tie_words:
+        length = len(word)
+        if length in word_lengths:
+            word_lengths[length].append(word)
+        else:
+            word_lengths[length] = [word]
```
Comment (mmcknett-ada) [inline adagrams/game.py:123]: [nit] This is an unusual approach! It would be great to add a comment explaining it. 😄

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,139 @@
...
+    # check a word has 10 characters
+
+    if 10 in word_lengths:
+        max_key_ten = word_lengths[10][0]
+        return max_key_ten, max_value
+    
+    # check shortest 
+
+    minimum_key = min(word_lengths)
+    print(minimum_key)
```
Comment (mmcknett-ada) [inline adagrams/game.py:134]: [nit] You can remove these debug print statements before submitting.

File: adagrams/game.py (hunk 5 above)
Comment (mmcknett-ada) [reply adagrams/game.py:123]: [suggestion] This implementation of `get_highest_word_score` is a little more complex than it could be.

Sometimes when we do these kinds of algorithms that have confusing layers of logical requirements, it's easy to get stuck in a corner by one early-on decision. In this case, I think you got cornered by trying to do this function with the built-in `max` and `min` functions. You'll probably find that if you start rewriting this function from scratch and begin with calculating "max" using a traditional for loop, you'll end up with a simpler, but still correct, result.

**Review by mmcknett-ada — COMMENTED** [review]
Looks great! 🟢 Green!

I have some nitpicks, comments, and suggestions, but those don't detract from the fact that this is a solid submission. Very nice!

Some overall feedback:

- The tests pass and your code even satisfies the corner cases I've found while grading that the tests don't cover.
- There is one missing requirement (the `E` 12x as likely as `Z` one), but the tests can't actually test for it. I recommend revisiting and thinking about how you might implement that, as well. (Most people miss this requirement, including me the first time I did the problem.)
- Your `get_highest_word_score` is a little complicated; rewriting it without `max` or `min` might actually give you a simpler result!
- I like that you did commits per wave and also that you committed as you were still working on waves. Committing as you solve sub-problems will continue to help you as a dev.
